### PR TITLE
Integrate mermaid in catalog

### LINKED ISSUE #13: Change diagrams to use mermaid js
Allow for either to be used

### PR DESCRIPTION
**Pull request by ashwinsajiv**
Fixes #13

### CODE REVIEW

File: pkg/catalog/diagram.go [hunk 1]
```diff
@@ -223,21 +259,28 @@ func (p *Generator) CreateReturnDataModel(appname string, stmnt *sysl.Statement,
...
-		plantumlString = catalogdiagrams.GenerateDataModel(appName, relatedTypes)
-		if _, ok := p.RootModule.GetApps()[appName]; !ok {
-			return ""
-		}
-		// Handle Empty
-		if len(relatedTypes) == 0 {
-			return ""
+	if !p.Mermaid {
+		var plantumlString string
+		if recursive {
```
Comment (joshcarp) [inline pkg/catalog/diagram.go]: This logic should be in a different function; eg split ` CreateTypeDiagram(` to delegate just like what was done with `CreateSequenceDiagram`

Comment (ashwinsajiv) [reply pkg/catalog/diagram.go]: Alright

Comment (ashwinsajiv) [reply pkg/catalog/diagram.go]: Addressed

**Comment by cuminandpaprika** [comment]
EPA Diagrams: 

Getting this mermaid output which looks correct to me:

But it doesn't seem to work in the live editor, is there some invalid mermaid syntax here? 

```
graph TD
 subgraph simple2
  FirstDivision-->ApplicationPackage
  SecondDivision-->MegaDatabase
  ThirdDivision-->ServerPackage
  ThirdDivision-->foo
  ThirdDivision-->simpleredoc
 end
 subgraph foo
  whatever-->...
 end
 subgraph this
  this-->...
 end
 subgraph that
  this-->this-this
 end
 subgraph RelModel
 end
 subgraph MobileApp
  Login-->Server-Authenticate
  Login-->ok_MegaDatabase.Empty
  Another-->ok_Server.Response
 end
 subgraph MobileApp2
  SecondEndpoint-->Server-Authenticate
  SecondEndpoint-->ok_MegaDatabase.Empty
  ThirdEndpoint-->ok_Server.Response
 end
 subgraph Server
  GET/testRestQueryParamPrimitive-->ok_Response
  GET/testRestPathParamPrimitive/_primitiveID_-->ok_Response
  GET/testReturnNil-->Empty
  Authenticate-->ok_sequenceofResponse
  GET/testRestURLParam/_id_-->ok_Response
  GET/testRestQueryParam-->ok_Response
 end
 subgraph MegaDatabase
 end
 subgraph simpleredoc
  GET/test-->error
  GET/test-->ok_SimpleObj
 end
 this-this-->this
 Server-Authenticate-->Authenticate
```

**Comment by cuminandpaprika** [comment]
Okay after some more testing, looks like a few more issues are afoot. Happy to look into these and push some fixes, just documenting them here so I don't forget tomorrow!

1. It looks like the SVGs aren't correctly being generated for some of the mermaid diagram types. Looks like some unit tests could to be added to `mermaid-go` isolate this or rule it out as a potential problem source.

To replicate, run `go run . -o demo/markdown demo/simple2.sysl --mermaid` and check the resulting file `demo/markdown/Server/request.svg`

The output mermaid string is:

```
classDiagram
 class Server_Response {
  MegaDatabase_Money query
  MegaDatabase_Empty balance
 }
 class Server_Request {
  List<Server_Response> query
 }
 Server_Request <-- Server_Response
```
And the SVG isn't displaying correctly for me
```
<svg id="mermaid-1592986859933" width="100%" xmlns="http://www.w3.org/2000/svg"><style>#mermaid-1592986859933 .label{font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family);color:#333}#mermaid-1592986859933 .label text{fill:#333}#mermaid-1592986859933 .node rect,#mermaid-1592986859933 .node circle,#mermaid-1592986859933 .node ellipse,#mermaid-1592986859933 .node polygon{fill:#ECECFF;stroke:#9370db;stroke-width:1px}#mermaid-1592986859933 .node .label{text-align:center}#mermaid-1592986859933 .node.clickable{cursor:pointer}#mermaid-1592986859933 .arrowheadPath{fill:#333}#mermaid-1592986859933 .edgePath .path{stroke:#333;stroke-width:1.5px}#mermaid-1592986859933 .edgeLabel{background-color:#e8e8e8;text-align:center}#mermaid-1592986859933 .cluster rect{fill:#ffffde;stroke:#aa3;stroke-width:1px}#mermaid-1592986859933 .cluster text{fill:#333}#mermaid-1592986859933 div.mermaidTooltip{position:absolute;text-align:center;max-width:200px;padding:2px;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family);font-size:12px;background:#ffffde;border:1px solid #aa3;border-radius:2px;pointer-events:none;z-index:100}#mermaid-1592986859933 .actor{stroke:#ccf;fill:#ECECFF}#mermaid-1592986859933 text.actor{fill:#000;stroke:none}#mermaid-1592986859933 .actor-line{stroke:grey}#mermaid-1592986859933 .messageLine0{stroke-width:1.5;stroke-dasharray:'2 2';stroke:#333}#mermaid-1592986859933 .messageLine1{stroke-width:1.5;stroke-dasharray:'2 2';stroke:#333}#mermaid-1592986859933 #arrowhead{fill:#333}#mermaid-1592986859933 .sequenceNumber{fill:#fff}#mermaid-1592986859933 #sequencenumber{fill:#333}#mermaid-1592986859933 #crosshead path{fill:#333 !important;stroke:#333 !important}#mermaid-1592986859933 .messageText{fill:#333;stroke:none}#mermaid-1592986859933 .labelBox{stroke:#ccf;fill:#ECECFF}#mermaid-1592986859933 .labelText{fill:#000;stroke:none}#mermaid-1592986859933 .loopText{fill:#000;stroke:none}#mermaid-1592986859933 .loopLine{stroke-width:2;stroke-dasharray:'2 2';stroke:#ccf}#mermaid-1592986859933 .note{stroke:#aa3;fill:#fff5ad}#mermaid-1592986859933 .noteText{fill:black;stroke:none;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family);font-size:14px}#mermaid-1592986859933 .activation0{fill:#f4f4f4;stroke:#666}#mermaid-1592986859933 .activation1{fill:#f4f4f4;stroke:#666}#mermaid-1592986859933 .activation2{fill:#f4f4f4;stroke:#666}#mermaid-1592986859933 .mermaid-main-font{font-family:"trebuchet ms", verdana, arial;font-family:var(--mermaid-font-family)}#mermaid-1592986859933 .section{stroke:none;opacity:0.2}#mermaid-1592986859933 .section0{fill:rgba(102,102,255,0.49)}#mermaid-1592986859933 .section2{fill:#fff400}#mermaid-1592986859933 .section1,#mermaid-1592986859933 .section3{fill:#fff;opacity:0.2}#mermaid-1592986859933 .sectionTitle0{fill:#333}#mermaid-1592986859933 .sectionTitle1{fill:#333}#mermaid-1592986859933 .sectionTitle2{fill:#333}#mermaid-1592986859933 .sectionTitle3{fill:#333}#mermaid-1592986859933 .sectionTitle{text-anchor:start;font-size:11px;text-height:14px;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family)}#mermaid-1592986859933 .grid .tick{stroke:#d3d3d3;opacity:0.3;shape-rendering:crispEdges}#mermaid-1592986859933 .grid .tick text{font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family)}#mermaid-1592986859933 .grid path{stroke-width:0}#mermaid-1592986859933 .today{fill:none;stroke:red;stroke-width:2px}#mermaid-1592986859933 .task{stroke-width:2}#mermaid-1592986859933 .taskText{text-anchor:middle;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family)}#mermaid-1592986859933 .taskText:not([font-size]){font-size:11px}#mermaid-1592986859933 .taskTextOutsideRight{fill:#000;text-anchor:start;font-size:11px;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family)}#mermaid-1592986859933 .taskTextOutsideLeft{fill:#000;text-anchor:end;font-size:11px}#mermaid-1592986859933 .task.clickable{cursor:pointer}#mermaid-1592986859933 .taskText.clickable{cursor:pointer;fill:#003163 !important;font-weight:bold}#mermaid-1592986859933 .taskTextOutsideLeft.clickable{cursor:pointer;fill:#003163 !important;font-weight:bold}#mermaid-1592986859933 .taskTextOutsideRight.clickable{cursor:pointer;fill:#003163 !important;font-weight:bold}#mermaid-1592986859933 .taskText0,#mermaid-1592986859933 .taskText1,#mermaid-1592986859933 .taskText2,#mermaid-1592986859933 .taskText3{fill:#fff}#mermaid-1592986859933 .task0,#mermaid-1592986859933 .task1,#mermaid-1592986859933 .task2,#mermaid-1592986859933 .task3{fill:#8a90dd;stroke:#534fbc}#mermaid-1592986859933 .taskTextOutside0,#mermaid-1592986859933 .taskTextOutside2{fill:#000}#mermaid-1592986859933 .taskTextOutside1,#mermaid-1592986859933 .taskTextOutside3{fill:#000}#mermaid-1592986859933 .active0,#mermaid-1592986859933 .active1,#mermaid-1592986859933 .active2,#mermaid-1592986859933 .active3{fill:#bfc7ff;stroke:#534fbc}#mermaid-1592986859933 .activeText0,#mermaid-1592986859933 .activeText1,#mermaid-1592986859933 .activeText2,#mermaid-1592986859933 .activeText3{fill:#000 !important}#mermaid-1592986859933 .done0,#mermaid-1592986859933 .done1,#mermaid-1592986859933 .done2,#mermaid-1592986859933 .done3{stroke:grey;fill:#d3d3d3;stroke-width:2}#mermaid-1592986859933 .doneText0,#mermaid-1592986859933 .doneText1,#mermaid-1592986859933 .doneText2,#mermaid-1592986859933 .doneText3{fill:#000 !important}#mermaid-1592986859933 .crit0,#mermaid-1592986859933 .crit1,#mermaid-1592986859933 .crit2,#mermaid-1592986859933 .crit3{stroke:#f88;fill:red;stroke-width:2}#mermaid-1592986859933 .activeCrit0,#mermaid-1592986859933 .activeCrit1,#mermaid-1592986859933 .activeCrit2,#mermaid-1592986859933 .activeCrit3{stroke:#f88;fill:#bfc7ff;stroke-width:2}#mermaid-1592986859933 .doneCrit0,#mermaid-1592986859933 .doneCrit1,#mermaid-1592986859933 .doneCrit2,#mermaid-1592986859933 .doneCrit3{stroke:#f88;fill:#d3d3d3;stroke-width:2;cursor:pointer;shape-rendering:crispEdges}#mermaid-1592986859933 .milestone{transform:rotate(45deg) scale(0.8, 0.8)}#mermaid-1592986859933 .milestoneText{font-style:italic}#mermaid-1592986859933 .doneCritText0,#mermaid-1592986859933 .doneCritText1,#mermaid-1592986859933 .doneCritText2,#mermaid-1592986859933 .doneCritText3{fill:#000 !important}#mermaid-1592986859933 .activeCritText0,#mermaid-1592986859933 .activeCritText1,#mermaid-1592986859933 .activeCritText2,#mermaid-1592986859933 .activeCritText3{fill:#000 !important}#mermaid-1592986859933 .titleText{text-anchor:middle;font-size:18px;fill:#000;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family)}#mermaid-1592986859933 g.classGroup text{fill:#9370db;stroke:none;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family);font-size:10px}#mermaid-1592986859933 g.classGroup text .title{font-weight:bolder}#mermaid-1592986859933 g.classGroup rect{fill:#ECECFF;stroke:#9370db}#mermaid-1592986859933 g.classGroup line{stroke:#9370db;stroke-width:1}#mermaid-1592986859933 .classLabel .box{stroke:none;stroke-width:0;fill:#ECECFF;opacity:0.5}#mermaid-1592986859933 .classLabel .label{fill:#9370db;font-size:10px}#mermaid-1592986859933 .relation{stroke:#9370db;stroke-width:1;fill:none}#mermaid-1592986859933 #compositionStart{fill:#9370db;stroke:#9370db;stroke-width:1}#mermaid-1592986859933 #compositionEnd{fill:#9370db;stroke:#9370db;stroke-width:1}#mermaid-1592986859933 #aggregationStart{fill:#ECECFF;stroke:#9370db;stroke-width:1}#mermaid-1592986859933 #aggregationEnd{fill:#ECECFF;stroke:#9370db;stroke-width:1}#mermaid-1592986859933 #dependencyStart{fill:#9370db;stroke:#9370db;stroke-width:1}#mermaid-1592986859933 #dependencyEnd{fill:#9370db;stroke:#9370db;stroke-width:1}#mermaid-1592986859933 #extensionStart{fill:#9370db;stroke:#9370db;stroke-width:1}#mermaid-1592986859933 #extensionEnd{fill:#9370db;stroke:#9370db;stroke-width:1}#mermaid-1592986859933 .commit-id,#mermaid-1592986859933 .commit-msg,#mermaid-1592986859933 .branch-label{fill:lightgrey;color:lightgrey;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family)}#mermaid-1592986859933 .pieTitleText{text-anchor:middle;font-size:25px;fill:#000;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family)}#mermaid-1592986859933 .slice{font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family)}#mermaid-1592986859933 g.stateGroup text{fill:#9370db;stroke:none;font-size:10px;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family)}#mermaid-1592986859933 g.stateGroup text{fill:#9370db;stroke:none;font-size:10px}#mermaid-1592986859933 g.stateGroup .state-title{font-weight:bolder;fill:#000}#mermaid-1592986859933 g.stateGroup rect{fill:#ECECFF;stroke:#9370db}#mermaid-1592986859933 g.stateGroup line{stroke:#9370db;stroke-width:1}#mermaid-1592986859933 .transition{stroke:#9370db;stroke-width:1;fill:none}#mermaid-1592986859933 .stateGroup .composit{fill:white;border-bottom:1px}#mermaid-1592986859933 .state-note{stroke:#aa3;fill:#fff5ad}#mermaid-1592986859933 .state-note text{fill:black;stroke:none;font-size:10px}#mermaid-1592986859933 .stateLabel .box{stroke:none;stroke-width:0;fill:#ECECFF;opacity:0.5}#mermaid-1592986859933 .stateLabel text{fill:#000;font-size:10px;font-weight:bold;font-family:'trebuchet ms', verdana, arial;font-family:var(--mermaid-font-family)}:root{--mermaid-font-family: '"trebuchet ms", verdana, arial';--mermaid-font-family: "Comic Sans MS", "Comic Sans", cursive}

:root { --mermaid-font-family: "trebuchet ms", verdana, arial;}</style><style>#mermaid-1592986859933 {
    color: rgb(0, 0, 0);
    font: 16px "trebuchet ms", verdana, arial;
  }</style><g></g></svg>
```

2. `GenerateAndWriteMermaidDiagram` in util.go is missing a statement for creating the directory in the fileName (since `build.sh` deletes the folders)
Adding something like this would solve it:
```
	if err := fs.MkdirAll(path.Dir(fileName), os.ModePerm); err != nil {
		return err
	}
```

3. The folder structure imposed by MacroPackages seems to be missing. You can see the results by running `go run . -o demo/markdown demo/simple2.sysl --mermaid` and looking at the diff between the generated files  in demo/markdown and the current committed files on the branch.

**Comment by cuminandpaprika** [comment]
PR taken over here: https://github.com/anz-bank/sysl-catalog/pull/139

**Pull request closed without merging**